src/memory_manager.cpp

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+#include "memory_manager.h"
+#include "utils.h"
+
+namespace diskann
+{
+
+size_t MemoryManager::get_memory_used_in_bytes() const
+{
+    return *_memory_used_in_bytes_ptr;
+}
+
+void MemoryManager::alloc_aligned(void **ptr, size_t size, size_t align)
+{
+    *ptr = nullptr;
+    if (IS_ALIGNED(size, align) == 0)
+        report_misalignment_of_requested_size(size, align);
+#ifndef _WINDOWS
+    *ptr = ::aligned_alloc(align, size);
+#else
+    *ptr = ::_aligned_malloc(size, align); // note the swapped arguments!
+#endif
+    if (*ptr == nullptr)
+    {
+        report_memory_allocation_failure(size, align);
+        return;
+    }
+
+    insert_block(*ptr, size);
+}
+
+void MemoryManager::realloc_aligned(void **ptr, size_t size, size_t align)
+{
+    if (IS_ALIGNED(size, align) == 0)
+        report_misalignment_of_requested_size(size, align);
+#ifdef _WINDOWS
+    erase_block(*ptr);
+    *ptr = ::_aligned_realloc(*ptr, size, align);
+#else
+    diskann::cerr << "No aligned realloc on GCC. Must malloc and mem_align, "
+                     "left it out for now."
+                  << std::endl;
+#endif
+    if (*ptr == nullptr)
+        report_memory_allocation_failure(size, align);
+
+    insert_block(*ptr, size);
+}
+
+void MemoryManager::aligned_free(void *ptr)
+{
+    // Gopal. Must have a check here if the pointer was actually allocated by
+    // _alloc_aligned
+    if (ptr == nullptr)
+    {
+        return;
+    }
+#ifndef _WINDOWS
+    free(ptr);
+#else
+    ::_aligned_free(ptr);
+#endif
+
+    erase_block(ptr);
+}
+
+void MemoryManager::insert_block(void *ptr, size_t size)
```

Comment:
this separate map for tracking allocation sizes seems very heavyweight, especially since it requires a hash-calculation for lookup in the map and locking to update the map.

What do you think of an approach of storing this info in the header of the memory block, like
buf = malloc(size + 4)
*buf = size
return buf + 4

I guess it requires some extra logic around alignment, but if we are okay with the possibility of a bit more padding waste I think it will be more performant.

Reply:
this is only used by alloc_aligned and new_array. Ideally, they are only used during loading index, they should not be used during searching index. so I expect this should not cause too much issue. But I am also ok to change to use the padding memory to store the size

Reply:
We need to make sure we are not allocating anything on the heap in insert or search functions. Otherwise, perf problems might arise.